### PR TITLE
chore: adjust for the Scenario 7.0 API

### PR DESCRIPTION
**Pull request by tonyandrewmeyer**
Updates the code to align with the Scenario 7.0 API changes - mostly the location of some imports and using sets in the e2e tests, but also handling the str to _Event conversion slightly differently.

This does expose `_Event` as an option to `run` even though it's now private. It seems like that's ok, and leaving the ability to use the string names here is most compatible. It is possible to do `CharmEvents.relation_changed(my_relation)` and so on (or `scenario.on.relation_changed(my_relation)` if we make that change in Scenario) if people need to customise the events or don't want to do string names.

### CODE REVIEW

File: interface_tester/interface_test.py [hunk 1]
```diff
@@ -408,21 +408,30 @@ def _run_scenario(self, event: Event, state: State):
...
+                raise InvalidTestCaseError(
+                    f"Bad interface test specification: event {raw_event} is not a relation event."
+                )
 
-        if not event._is_relation_event:
+        if not isinstance(event, _Event):
             raise InvalidTestCaseError(
-                f"Bad interface test specification: event {raw_event} " "is not a relation event."
+                f"Expected _Event or str, not {type(raw_event)}. "
+                f"Invalid test case: {self} cannot cast {raw_event} to _Event."
```
Comment (tonyandrewmeyer) [inline interface_tester/interface_test.py:434]:
```suggestion
                f"Bad interface test specification: event {raw_event} should be a relation event."
```

File: interface_tester/interface_test.py [hunk 2]
```diff
@@ -408,21 +408,29 @@ def _run_scenario(self, event: Event, state: State):
...
+            else:
+                raise InvalidTestCaseError(
+                    f"Bad interface test specification: event {raw_event} is not a relation event."
+                )
 
-        if not event._is_relation_event:
+        if not isinstance(event, _Event):
             raise InvalidTestCaseError(
-                f"Bad interface test specification: event {raw_event} " "is not a relation event."
+                f"Bad interface test specification: event {raw_event} should be a relation event."
```
Comment (tonyandrewmeyer) [inline interface_tester/interface_test.py:433]:
```suggestion
                f"Bad interface test specification: event {raw_event} should be a relation event string or _Event."
```

**Comment by tonyandrewmeyer** [comment]
@PietroPasotti could you give this a review when you have some time? This blocks a few charms from upgrading to Scenario 7, so would be good to get out of the way. Let me know if you'd rather I took a different approach.

File: interface_tester/interface_test.py [hunk 3]
```diff
@@ -379,22 +380,21 @@ def _run(self, event: Union[str, Event]):
         # some required config, a "happy" status, network information, OTHER relations.
         # Typically, should NOT touch the relation that this interface test is about
         #  -> so we overwrite and warn on conflict: state_template is the baseline,
-        state = (self.ctx.state_template or State()).copy()
+        state = copy.deepcopy(self.ctx.state_template or State())
```
Comment (PietroPasotti) [inline interface_tester/interface_test.py:383]: not sure if it's better but I usually do a `dataclasses.replace(dc_in)` instead of deepcopy, also it's pointless to copy if it's a fresh instance (not your fault, I know...)
```suggestion
        state = dataclasses.replace(self.ctx.state_template) if self.ctx.state_template else State()
```

File: interface_tester/interface_test.py [hunk 4]
```diff
@@ -379,22 +380,21 @@ def _run(self, event: Union[str, Event]):
...
-        modified_state = state.replace(relations=relations)
+        modified_state = dataclasses.replace(state, relations=relations)
 
         # the Relation instance this test is about:
         relation = next(filter(lambda r: r.interface == self.ctx.interface_name, relations))
-        evt: Event = self._cast_event(event, relation)
 
-        logger.info("collected test for %s with %s" % (self.ctx.interface_name, evt.name))
-        return self._run_scenario(evt, modified_state)
+        logger.info("collected test for %s with %s" % (self.ctx.interface_name, event))
```
Comment (PietroPasotti) [inline interface_tester/interface_test.py:394]: why this change?

Comment (tonyandrewmeyer) [reply interface_tester/interface_test.py:394]: I moved the `_cast_event` call into `_run_scenario` (I don't remember exactly why - I could figure that out if you like, or try to avoid it). Because of that, at this point `event` is `str|_Event` so the `.name` may not work.

I could also just move the logging line to be in `_run_scenario` also.

Comment (PietroPasotti) [reply interface_tester/interface_test.py:394]: uh. Well in general I try to coerce types as soon as possible to keep the private signatures nice and simple, so if it's not too much trouble...

Comment (tonyandrewmeyer) [reply interface_tester/interface_test.py]: Sure. It was to get access to `ctx.on`, but I've changed it to use `CharmEvents` directly instead, so it's much more minimal.

File: interface_tester/interface_test.py [hunk 5]
```diff
@@ -410,19 +413,27 @@ def _run_scenario(self, event: Event, state: State):
...
+                event = CharmEvents.relation_joined(relation)
+            elif raw_event.endswith("-relation-created"):
+                event = CharmEvents.relation_created(relation)
+            else:
+                raise InvalidTestCaseError(
+                    f"Bad interface test specification: event {raw_event} is not a relation event."
+                )
 
-        if not event._is_relation_event:
+        if not isinstance(event, _Event):
```
Comment (PietroPasotti) [inline interface_tester/interface_test.py:433]: ehm, what is "event" now referring to?

**Review by PietroPasotti — CHANGES_REQUESTED** [review]

File: interface_tester/interface_test.py (hunk 5 above)
Comment (PietroPasotti) [reply interface_tester/interface_test.py:433]: and why don't we do this type checking earlier?

Comment (tonyandrewmeyer) [reply interface_tester/interface_test.py]: > ehm, what is "event" now referring to?

Yeah, missed fixing the name there, sorry. Fixed now.

> and why don't we do this type checking earlier?

It felt cleaner to do the "if string then do this" first and then "if not _Event error" rather than "if not str or _Event error" followed by "if string". I've changed it back to the other order.

**Comment by tonyandrewmeyer** [comment]
@PietroPasotti are you ok with this as it is now?

**Review by PietroPasotti — APPROVED** [review]
yes! thanks

**Comment by tonyandrewmeyer** [comment]
> yes! thanks

Could you merge it for me, please?
